p4-16/psa/PSA.mdk

```diff
@@ -2473,6 +2473,56 @@ a common P4 program.
 We also need to formalize the interaction of action profiles and action selectors with counters and
 meters.
 
+## Observation and control of congestion
+
+The current PSA does not provide any mechanisms to observe if
+particular output ports or queues are leading to congestion in the
+packet buffer.  Thus it is not possible without using mechanisms
+defined outside of PSA to implement a feature like Explicit Congestion
+Notification (ECN)[^ECN].  One possibility here is to define a small
+field, perhaps only 1 bit, that is part of the metadata associated
+with each packet as it begins egress processing.  This field would
+indicate "how much congestion" the packet experienced in the packet
+buffer.
+
+[^ECN]: https://en.wikipedia.org/wiki/Explicit_Congestion_Notification
+
+There is also currently no way defined in PSA for ingress P4 code to
+send information about a packet to the packet buffer that might
+influence the behavior of a congestion control algorithm, such as
+Approximate Fair Drop (AFD).  This is partly because of the variety of
+congestion control mechanisms in use by switches today.
+
+It would be desirable to define in PSA a small set of fields about a
+packet that would be useful inputs to multiple congestion control
+algorithms.  One possibility is a hash of the packet's "flow id",
+often implemented as a hash of packet header fields like IP source and
+destination address, IP protocol, and optionally TCP/UDP source and
+destination ports.  Given that P4 programmable devices can implement
+network protocols other than IP, including custom ones, a more general
+mechanism is desirable in PSA devices.
+
+
+## Observation and control of congestion
```

Comment:
Cut and paste error for the title? Do you want to rename this to "Observing packet ordering"?

Reply:
This sub-section only deals with observing whether a packet experienced congestion, e.g. for implementing ECN header modifications in egress, or potentially feeding inputs to a congestion control algorithm from ingress.  I don't see any connection to packet ordering here.

I know there is an open issue about potentially adding ways in the future to tell the compiler restrictions it should follow on packet ordering for a particular P4 program.  I did not attempt to create an open issue in the document for that, mainly because it seemed like something fewer people would care about (not 0, but fewer).

Reply:
You have the same heading for both ECN and timestamps.

Reply:
Woops, disregard my previous comment.  Yes, this sub-section title is definitely wrong.  Fixing...